src/Chapter4.hs

```diff
@@ -471,11 +481,16 @@ Applicatives can be found in many applications:
 Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
-    pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+  pure :: a -> Secret e a
+  pure = Reward
 
-    (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+  (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
+  Trap f <*> _ = Trap f
+  Reward f <*> x        = fmap f x
+
+-- Trap f      <*> (Trap e) = Trap f
+-- this previosu implementation was breaking a case, 
+--  in Applicatives do we need to give precednece to the function ?
```

Comment:
What do you mean by "precedence to the function"? Sorry, I didn't quite get what was the previous implementation and how it failed...

Reply:
Oh Sorry,
actual implementation was 

 `Trap _    <*> (Trap e) = Trap e`

The test case failing:: 
`it "trap <*> trap" $
            Trap "it's a trap" <*> Trap "42" `shouldBe` trap`

The first param I felt needed to be a function under same context.

Thus I made that conclusion